File: Demos/VB.Net/DataFromBusinessObject/Form1.vb
Public Class Form1
    Private FBusinessObject As List(Of Category)

    Private Sub CreateBusinessObject()
        FBusinessObject = New List(Of Category)

        Dim category As New Category("Beverages", "Soft drinks, coffees, teas, beers")
        category.Products.Add(New Product("Chai", 18))
        category.Products.Add(New Product("Chang", 19))
        category.Products.Add(New Product("Ipoh coffee", 46))
        FBusinessObject.Add(category)

        category = New Category("Confections", "Desserts, candies, and sweet breads")
        category.Products.Add(New Product("Chocolade", 12.75))
        category.Products.Add(New Product("Scottish Longbreads", 12.5))
        category.Products.Add(New Product("Tarte au sucre", 49.3))
        FBusinessObject.Add(category)

        category = New Category("Seafood", "Seaweed and fish")
        category.Products.Add(New Product("Boston Crab Meat", 18.4))
        category.Products.Add(New Product("Red caviar", 15))
        FBusinessObject.Add(category)
    End Sub

    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        CreateBusinessObject()
    End Sub

    Private Sub btnCreateNew_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCreateNew.Click
        ' create report instance
        Dim report As New Report
        ' register the business object
        report.RegisterData(Me.FBusinessObject, "Categories")
        ' design the report
        report.Design()
        ' free resources used by report
        report.Dispose()
    End Sub

    Private Sub btnRunExisting_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnRunExisting.Click
        ' create report instance
        Dim report As New Report
        ' load the existing report
        report.Load("..\..\report.frx")
        ' register the business object
        report.RegisterData(Me.FBusinessObject, "Categories")
        ' show the report
        report.Show()
        ' free resources used by report
        report.Dispose()
    End Sub

    Public Class Category
        ' Fields
        Private FDescription As String
        Private FName As String
        Private FProducts As List(Of Product)

        ' Properties
        Public ReadOnly Property Description() As String
            Get
                Return Me.FDescription
            End Get
        End Property

        Public ReadOnly Property Name() As String
            Get
                Return Me.FName
            End Get
        End Property

        Public ReadOnly Property Products() As List(Of Product)
            Get
                Return Me.FProducts
            End Get
        End Property

        ' Methods
        Public Sub New(ByVal name As String, ByVal description As String)
            Me.FName = name
            Me.FDescription = description
            Me.FProducts = New List(Of Product)
        End Sub
    End Class

    Public Class Product
        ' Fields
        Private FName As String
        Private FUnitPrice As Decimal

        ' Properties
        Public ReadOnly Property Name() As String
            Get
                Return Me.FName
            End Get
        End Property

        Public ReadOnly Property UnitPrice() As Decimal
            Get
                Return Me.FUnitPrice
            End Get
        End Property

        ' Methods
        Public Sub New(ByVal name As String, ByVal unitPrice As Decimal)
            Me.FName = name
            Me.FUnitPrice = unitPrice
        End Sub
    End Class


End Class

File: Demos/C#/DataFromBusinessObject/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using FastReport;

namespace DataFromBusinessObject
{
  public partial class Form1 : Form
  {
    private List<Category> FBusinessObject;
    
    public Form1()
    {
      InitializeComponent();
      CreateBusinessObject();
    }

    private void CreateBusinessObject()
    {
      FBusinessObject = new List<Category>();

      Category category = new Category("Beverages", "Soft drinks, coffees, teas, beers");
      category.Products.Add(new Product("Chai", 18m));
      category.Products.Add(new Product("Chang", 19m));
      category.Products.Add(new Product("Ipoh coffee", 46m));
      FBusinessObject.Add(category);

      category = new Category("Confections", "Desserts, candies, and sweet breads");
      category.Products.Add(new Product("Chocolade", 12.75m));
      category.Products.Add(new Product("Scottish Longbreads", 12.5m));
      category.Products.Add(new Product("Tarte au sucre", 49.3m));
      FBusinessObject.Add(category);

      category = new Category("Seafood", "Seaweed and fish");
      category.Products.Add(new Product("Boston Crab Meat", 18.4m));
      category.Products.Add(new Product("Red caviar", 15m));
      FBusinessObject.Add(category);
    }

    private void btnCreateNew_Click(object sender, EventArgs e)
    {
      // create report instance
      Report report = new Report();

      // register the business object
      report.RegisterData(FBusinessObject, "Categories");

      // design the report
      report.Design();

      // free resources used by report
      report.Dispose();
    }

    private void btnRunExisting_Click(object sender, EventArgs e)
    {
      // create report instance
      Report report = new Report();

      // load the existing report
      report.Load(@"..\..\report.frx");

      // register the business object
      report.RegisterData(FBusinessObject, "Categories");

      // run the report
      report.Show();

      // free resources used by report
      report.Dispose();
    }
  }


  public class Category
  {
    private string FName;
    private string FDescription;
    private List<Product> FProducts;

    public string Name
    {
      get { return FName; }
    }

    public string Description
    {
      get { return FDescription; }
    }

    public List<Product> Products
    {
      get { return FProducts; }
    }

    public Category(string name, string description)
    {
      FName = name;
      FDescription = description;
      FProducts = new List<Product>();
    }
  }

  public class Product
  {
    private string FName;
    private decimal FUnitPrice;

    public string Name
    {
      get { return FName; }
    }

    public decimal UnitPrice
    {
      get { return FUnitPrice; }
    }

    public Product(string name, decimal unitPrice)
    {
      FName = name;
      FUnitPrice = unitPrice;
    }
  }
}